File: sirius-core/src/main/java/org/starcoin/sirius/core/ParticipantGang.java
package org.starcoin.sirius.core;

import com.google.protobuf.ByteString;
import org.starcoin.core.ProtobufCodec;
import org.starcoin.proto.Starcoin.ProtoParticipantGang;
import org.starcoin.util.KeyPairUtil;

import java.security.KeyPair;
import java.security.PrivateKey;

// just for test, mock malicious Participant
public class ParticipantGang implements ProtobufCodec<ProtoParticipantGang> {

    private Participant participant;
    private PrivateKey privateKey;

    public ParticipantGang() {
    }

    public ParticipantGang(Participant participant, PrivateKey privateKey) {
        this.participant = participant;
        this.privateKey = privateKey;
    }

    public Participant getParticipant() {
        return participant;
    }

    public PrivateKey getPrivateKey() {
        return privateKey;
    }

    @Override
    public ProtoParticipantGang marshalProto() {
        return ProtoParticipantGang.newBuilder()
                .setParticipant(this.participant.toProto())
                .setPrivateKey(ByteString.copyFrom(KeyPairUtil.encodePrivateKey(this.privateKey)))
                .build();
    }

    @Override
    public void unmarshalProto(ProtoParticipantGang proto) {
        this.participant = proto.hasParticipant() ? new Participant(proto.getParticipant()) : null;
        this.privateKey =
                proto.getPrivateKey().isEmpty()
                        ? null
                        : KeyPairUtil.recoverPrivateKey(proto.getPrivateKey().toByteArray());
    }

    public static ParticipantGang random() {
        KeyPair keyPair = KeyPairUtil.generateKeyPair();
        return new ParticipantGang(new Participant(keyPair.getPublic()), keyPair.getPrivate());
    }
}

File: sirius-core/src/main/kotlin/org/starcoin/sirius/core/ParticipantGang.kt
package org.starcoin.sirius.core

import kotlinx.serialization.SerialId
import kotlinx.serialization.Serializable
import org.starcoin.proto.Starcoin
import org.starcoin.sirius.crypto.CryptoService
import org.starcoin.sirius.serialization.PrivateKeySerializer
import org.starcoin.sirius.serialization.ProtobufSchema
import java.security.PrivateKey

// just for test, mock malicious Participant
@ProtobufSchema(Starcoin.ProtoParticipantGang::class)
@Serializable
data class ParticipantGang(
    @SerialId(1)
    val participant: Participant,
    @SerialId(2)
    @Serializable(with = PrivateKeySerializer::class)
    val privateKey: PrivateKey
) : SiriusObject() {


    companion object : SiriusObjectCompanion<ParticipantGang, Starcoin.ProtoParticipantGang>(ParticipantGang::class) {

        override fun mock(): ParticipantGang {
            return random()
        }

        fun random(): ParticipantGang {
            val key = CryptoService.generateCryptoKey()
            return ParticipantGang(Participant(key.keyPair.public), key.keyPair.private)
        }
    }
}
